extensions/sftp/processors/ListSFTP.cpp

```diff
@@ -0,0 +1,1320 @@
+      } else {
+        logger_->log_debug("Precision auto detection detected minute precision");
+        remote_system_timestamp_precision = TARGET_SYSTEM_TIMESTAMP_PRECISION_MINUTES;
+      }
+    } else if (target_system_timestamp_precision_ == TARGET_SYSTEM_TIMESTAMP_PRECISION_MINUTES) {
+        remote_system_timestamp_precision = TARGET_SYSTEM_TIMESTAMP_PRECISION_MINUTES;
+    } else {
+      /*
+       * We only have seconds-precision timestamps, TARGET_SYSTEM_TIMESTAMP_PRECISION_MILLISECONDS makes no real sense here,
+       * so we will treat it as TARGET_SYSTEM_TIMESTAMP_PRECISION_SECONDS.
+       */
+      remote_system_timestamp_precision = TARGET_SYSTEM_TIMESTAMP_PRECISION_SECONDS;
+    }
+    uint64_t listing_lag = LISTING_LAG_MAP.at(remote_system_timestamp_precision);
+    logger_->log_debug("The listing lag is %lu ms", listing_lag);
+
+    /* If the latest listing time is equal to the last listing time, there are no entries with a newer timestamp than previously seen */
+    if (latest_listed_entry_timestamp_this_cycle == last_listed_latest_entry_timestamp_) {
+      const auto& latest_files = ordered_files.at(latest_listed_entry_timestamp_this_cycle);
+      uint64_t elapsed_time = std::chrono::duration_cast<std::chrono::milliseconds>(current_run_time - last_run_time_).count();
+      /* If a precision-specific listing lag has not yet elapsed since out last execution, we wait. */
+      if (elapsed_time < listing_lag) {
+        logger_->log_debug("The latest listed entry timestamp is the same as the last listed entry timestamp (%lu) "
+                           "and the listing lag has not yet elapsed (%lu ms < % lu ms). Yielding.",
+                           latest_listed_entry_timestamp_this_cycle,
+                           elapsed_time,
+                           listing_lag);
+        context->yield();
+        return;
+      }
+      /*
+       * If we have already processed the entities with the newest timestamp,
+       * and there are no new entities with that timestamp, there is nothing to do.
+       */
+      if (latest_listed_entry_timestamp_this_cycle == last_processed_latest_entry_timestamp_ &&
+          std::all_of(latest_files.begin(), latest_files.end(), [this](const Child& child) {
+            return latest_identifiers_processed_.count(child.getPath()) == 1U;
+          })) {
+        logger_->log_debug("The latest listed entry timestamp is the same as the last listed entry timestamp (%lu) "
+                           "and all files for that timestamp has been processed. Yielding.", latest_listed_entry_timestamp_this_cycle);
+        context->yield();
+        return;
+      }
+    } else {
+      /* Determine the minimum reliable timestamp based on precision */
+      uint64_t minimum_reliable_timestamp = now * 1000 - listing_lag;
+      if (remote_system_timestamp_precision == TARGET_SYSTEM_TIMESTAMP_PRECISION_SECONDS) {
```

Comment:
I feel there will be a time soon where this logic will have to be generalized to be used with other sources as well (FTP, HDFS, etc.).
When that time comes TARGET_SYSTEM_TIMESTAMP_PRECISION_MILLISECONDS can be a valid option, which would require us to create a nasty ternary-conditional chain, which I like to avoid, so I would keep this as it is for now.

Reply:
That makes sense. 
In case it gets complex, I think the appropriate way of doing so is using std::chrono::duration, which supports all of these and provides a "floor" method, too.

Reply:
You are right, many parts of the code (this not being an exception) uses unix timestamps in milliseconds stored in uint64_t's/int64_t's. It would be better to use a proper type system for timestamps and durations everywhere.